crossbeam-epoch/src/atomic.rs

```diff
@@ -109,22 +110,56 @@ fn low_bits<T: ?Sized + Pointable>() -> usize {
 
 /// Panics if the pointer is not properly unaligned.
 #[inline]
-fn ensure_aligned<T: ?Sized + Pointable>(raw: usize) {
-    assert_eq!(raw & low_bits::<T>(), 0, "unaligned pointer");
+fn ensure_aligned<T: ?Sized + Pointable>(raw: *mut ()) {
+    assert_eq!(raw as usize & low_bits::<T>(), 0, "unaligned pointer");
 }
 
 /// Given a tagged pointer `data`, returns the same pointer, but tagged with `tag`.
 ///
 /// `tag` is truncated to fit into the unused bits of the pointer to `T`.
 #[inline]
-fn compose_tag<T: ?Sized + Pointable>(data: usize, tag: usize) -> usize {
-    (data & !low_bits::<T>()) | (tag & low_bits::<T>())
+fn compose_tag<T: ?Sized + Pointable>(ptr: *mut (), tag: usize) -> *mut () {
+    int_to_ptr_with_provenance(
+        (ptr as usize & !low_bits::<T>()) | (tag & low_bits::<T>()),
+        ptr,
+    )
 }
 
 /// Decomposes a tagged pointer `data` into the pointer and the tag.
 #[inline]
-fn decompose_tag<T: ?Sized + Pointable>(data: usize) -> (usize, usize) {
-    (data & !low_bits::<T>(), data & low_bits::<T>())
+fn decompose_tag<T: ?Sized + Pointable>(ptr: *mut ()) -> (*mut (), usize) {
+    (
+        int_to_ptr_with_provenance(ptr as usize & !low_bits::<T>(), ptr),
+        ptr as usize & low_bits::<T>(),
+    )
+}
+
+// HACK: https://github.com/rust-lang/miri/issues/1866#issuecomment-985802751
+#[inline]
+fn int_to_ptr_with_provenance<T>(addr: usize, prov: *mut T) -> *mut T {
+    let ptr = prov as *mut u8;
+    ptr.wrapping_add(addr.wrapping_sub(ptr as usize)) as *mut T
+}
```

Comment:
This hack will likely be unnecessary in the future: https://github.com/rust-lang/rust/pull/96935